### ISSUE TITLE
Enhancement: [prefer-for-of] recognize memoized array.length

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#4979]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-for-of

### Description

Context: https://github.com/typescript-eslint/typescript-eslint/pull/4977#discussion_r873167767

Memoizing `array.length` ahead-of-time is a very common optimization technique.

### Fail

```typescript
for (let i = 0, l = unusedVars.length; i < l; ++i) {
  console.log(unusedVars[i]);
}
```

### Pass

```typescript
for (const unusedVar of unusedVars) {
  console.log(unusedVar);
}
```

**Comment by armano2** [comment]
this is really good idea, it may be a little complex to detect

**Comment by bradzacher** [comment]
Off topic, but to clarify

> Memoizing array.length ahead-of-time is a very common optimization technique

A quick google shows that the lookup is optimised in modern browsers to be a constant value lookup (I.e. You're talking maybe a few CPU ops of difference between that and a local var lookup).

This is a micro-optimisation that will likely shave off a few ***micro***seconds, if that. IMO nobody should make this optimisation because it makes the code so much worse to read.

**Comment by Josh-Cena** [comment]
Apparently it's used in our own code😄

Also, folks seem to prefer iterating backwards (`for (let i = arr.length - 1; i >=0; i--)`) in hope that it optimizes length lookup.

**Comment by bradzacher** [comment]
It probably is - and really it should be linted against by itself.

Ive seen a lot of weird code that people copy paste from other languages around loops. I've even seen this before:
```ts
var i;
for (i = 0; ...) { ... }
for (i = 0; ...) { ... }
... 
```
Reusing a loop iteration variable to optimise and save a declaration...

People do write bad code!

**Comment by bradzacher** [comment]
Based on our current expectations for enhancement requests, I believe this request fails the "nicheness" test:
- This request has been open for 4 years and has netted a total of 0 non-team member interactions.
- This request has been open for 4 years and nobody has moved to implement it.
- The programming style in question is niche / non-standard.

For these reasons, I believe we should move to reject and close this. Thoughts @typescript-eslint/triage-team?

**Comment by Josh-Cena** [comment]
Agreed, I think it's getting less relevant.
